### PR TITLE
Flash Data Sets - Detect Custom Fields

### PR DESCRIPTION
**Pull request by cbyrohl**
## PR Summary

For FLASH data sets, currently only regular plot variables (e.g. "plot_var_1",..., [see here](http://flash.uchicago.edu/site/flashcode/user_support/flash_ug_devel/node68.html)) are detected. However, FLASH also has additional, user-defined scratch variables ("plot_grid_var_1",..., [see here](http://flash.uchicago.edu/site/flashcode/user_support/flash_ug_devel/node72.html)). The PR adds the later fields to the field list to be used in yt.

Unclear whether this needs to be documented somewhere +  whether a test for this feature is to be included as it would require a data set with such custom field.

## PR Checklist

- [x] pass `flake8 yt/`
- [x] pass `isort -rc . --check-only`
- [x] pass `black --check yt/`
- [ ] New features are documented, with docstrings and narrative docs
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Looks good to me, thank you for this update !
It would indeed be useful to add tests for this but the process isn't very straightforward, and I think you'd need to contribute a dataset with these new fields to our database; Here's how to do it:

1 - open a PR to https://github.com/yt-project/website to propose your dataset (including a dl link), register and describe it dataset in this file:
https://github.com/yt-project/website/blob/master/data/datafiles.json
2 - add a tests to `yt/frontends/flash/tests/` that you'll need to decorate with `@requires_ds()` so that it only runs on machines where the dataset is present, this can be done in the present PR.

note that it will be looked for in the test data file, which you need configure in yt's config file in order to run your test locally (see https://yt-project.org/doc/reference/configuration.html#the-configuration-file), and the file/dir needs to be named as referenced by `datafiles.json` ("filename").

Let us know if you need help with this, or anything is still not clear (I know I didn't get it on first try !)

File: yt/frontends/flash/data_structures.py [hunk 1]
```diff
@@ -53,6 +53,15 @@ def _detect_output_fields(self):
...
             for s in self._handle["/unknown names"][:].flat
         ]
+        # add custom/scratch fields:
+        srp = self._handle["/string runtime parameters"]
+        for i in range(srp.shape[0]):
+            k = srp[i][0]
+            if k.startswith(b"plot_grid_var"):
+                v = srp[i][1].decode("ascii", "ignore")[:4]
+                if v != "none":
+                    self.field_list.append(("flash", v))
```
Comment (neutrinoceros) [inline yt/frontends/flash/data_structures.py:63]:
```suggestion
        for k, v in srp:
            fname = v.decode("ascii", "ignore")[:4]
            if k.startswith(b"plot_grid_var") and fname != "none":
                self.field_list.append(("flash", fname))
```

Comment (matthewturk) [reply yt/frontends/flash/data_structures.py:63]: This looks like an improvement over the original *as long as* we are sure that iterating over `srp` in the way you propose still works with the way `string runtime parameters` are stored.

Comment (neutrinoceros) [reply yt/frontends/flash/data_structures.py:63]: of course ! I'm taking a bet here :)

**Comment by neutrinoceros** [comment]
Hey @cbyrohl, the failing tests seen here have been resolved on master, can you please update your branch to complete the validation ? Thanks !

**Comment by munkm** [comment]
Hi @cbyrohl I thought I'd follow up with you on this. Do you have any datasets with these types of fields that you're willing to share? One of us can help add the data to the website (and our testing suite) to support this PR.

File: yt/frontends/flash/data_structures.py (hunk 1 above)
Comment (neutrinoceros) [reply yt/frontends/flash/data_structures.py:63]: @cbyrohl, what's your thinking on this ? :-)

**Comment by cbyrohl** [comment]
Hi @munkm , let me try to provide such data set and according test towards the end of this week. My current concern is that the Flash code might have changed some of its behavior in recent release as the code base seemed to have changed for "user grid vars" quite a bit. I do not have time to check this, the data set would be based on Flash 4.3.

Best,
Chris

**Comment by munkm** [comment]
I think that's totally fine! If your PR works for a particular version of flash, then we can open issues down the road if Flash behavior changes with new updates! You've added the important first-step infrastructure anyways. 🙂 🎉 

Thank you so much for this! If you send the file to me I'll add it to the website.

**Comment by matthewturk** [comment]
Hi!  We'd like to help get your work in -- it looks like we're pretty close to the finish line, @cbyrohl .  Anything we can do to help out?

**Comment by munkm** [comment]
Hi @cbyrohl we'd love to get this feature into yt! Is there any help we can provide to help you out?

**Comment by neutrinoceros** [comment]
@cbyrohl do you still wish to get this patch merged ? I know it's been a while but I see your account is still active so I'm reaching out one more time. If this gets no response in the following weeks I'd vote to close this PR as stale.

**Comment by matthewturk** [comment]
I'd be happy to help if you need anything.

**Comment by cbyrohl** [comment]
Apologies, as I am not working with yt/FLASH these days and have not set this up on the computing resources currently available to me right now, I cannot finish this pull request at this time.

**Comment by neutrinoceros** [comment]
Thank you for responding. We would normally offer take over, unfortunately it seems that the FLASH source code and documentation are currently unavailable (been trying to access them for a week), so it's a bit hard to pursue this work in the blind. Given the situation I'm going to close this PR. I'm sorry this didn't come through.
For posterity: we would still happily consider a similar patch for inclusion, hopefully we'll do better next time.

**Comment by jzuhone** [comment]
@neutrinoceros FLASH has moved to the University of Rochester:

flash.rochester.edu

**Comment by neutrinoceros** [comment]
Great, thanks !
Would you be up for taking over this ? Feel free to reopen regardless.

**Pull request closed without merging**